compiler/src/dotty/tools/dotc/typer/Checking.scala

```diff
@@ -594,6 +597,48 @@ trait Checking {
       ctx.error(ex"$cls cannot be instantiated since it${rstatus.msg}", pos)
   }
 
+  /** Check that pattern `pat` is irrefutable for scrutinee tye `pt`.
+   *  This means `pat` is either marked @unchecked or `pt` conforms to the
+   *  pattern's type. If pattern is an UnApply, do the check recursively.
+   */
+  def checkIrrefutable(pat: Tree, pt: Type)(implicit ctx: Context): Boolean = {
+    patmatch.println(i"check irrefutable $pat: ${pat.tpe} against $pt")
+
+    def check(pat: Tree, pt: Type): Boolean = {
+      if (pt <:< pat.tpe)
+        true
+      else {
+        ctx.errorOrMigrationWarning(
+          ex"""pattern's type ${pat.tpe} is more specialized than the right hand side expression's type ${pt.dropAnnot(defn.UncheckedAnnot)}
+              |
+              |If the narrowing is intentional, this can be communicated by writing `: @unchecked` after the full pattern.${err.rewriteNotice}""",
+          pat.sourcePos)
+        false
+      }
+    }
+
+    !ctx.settings.strict.value || // only in -strict mode for now since mitigations work only after this PR
+    pat.tpe.widen.hasAnnotation(defn.UncheckedAnnot) || {
+      pat match {
+        case Bind(_, pat1) =>
+          checkIrrefutable(pat1, pt)
+        case UnApply(fn, _, pats) =>
+          check(pat, pt) && {
+            val argPts = unapplyArgs(fn.tpe.finalResultType, fn, pats, pat.sourcePos)
+            pats.corresponds(argPts)(checkIrrefutable)
+          }
```

Comment:
For `UnApply`, we should also check if the extractor itself is irrefutable, i.e. if it returns `Some`, or `true`, or a product. The logic to do that is in `SpaceEngine#irrefutable`.

Additionally, for some reason unapplies are simply rejected as `val` definitions:
```
scala> {
     |   object Positive { def unapply(i: Int): Option[Int] = Some(i).filter(_ > 0) }
     |   val Positive(p) = 5
     |   5 match { case Positive(p) => p }
     | }
3 |  val Positive(p) = 5
  |      ^^^^^^^^^^^
  | ((i: Int): Option[Int])(Positive.unapply) is not a valid result type of an unapply method of an extractor.
```